crypto/bn/asm/rsaz-2k-avxifma.pl

```diff
@@ -0,0 +1,863 @@
+#                                       const BN_ULONG *a,
+#                                       const BN_ULONG *b,
+#                                       const BN_ULONG *m,
+#                                       BN_ULONG k0);
+###############################################################################
+{
+# input parameters ("%rdi","%rsi","%rdx","%rcx","%r8")
+my ($res,$a,$b,$m,$k0) = @_6_args_universal_ABI;
+
+my $mask52     = "%rax";
+my $acc0_0     = "%r9";
+my $acc0_0_low = "%r9d";
+my $acc0_1     = "%r15";
+my $acc0_1_low = "%r15d";
+my $b_ptr      = "%r11";
+
+my $iter = "%ebx";
+
+my $zero = "%ymm0";
+my $Bi   = "%ymm1";
+my $Yi   = "%ymm2";
+my $Yi_xmm   = "%xmm2";
+my ($R0_0,$R0_0h,$R1_0,$R1_0h,$R2_0) = ("%ymm3",map("%ymm$_",(5..8)));
+my ($R0_1,$R0_1h,$R1_1,$R1_1h,$R2_1) = ("%ymm4",map("%ymm$_",(9..12)));
+
+# Registers mapping for normalization.
+my ($T0,$T0h,$T1,$T1h,$T2,$tmp) = ("$zero", "$Bi", "$Yi", map("%ymm$_", (13..15)));
+my $T0_xmm   = "%xmm0";
+
+sub amm52x20_x1() {
+# _data_offset - offset in the |a| or |m| arrays pointing to the beginning
+#                of data for corresponding AMM operation;
+# _b_offset    - offset in the |b| array pointing to the next qword digit;
+my ($_data_offset,$_b_offset,$_acc,$_R0,$_R0h,$_R1,$_R1h,$_R2,$_k0) = @_;
+$code.=<<___;
+    movq    $_b_offset($b_ptr), %r13             # b[i]
+
+    vpbroadcastq    $_b_offset($b_ptr), $Bi      # broadcast b[i]
+    movq    $_data_offset($a), %rdx
+    mulx    %r13, %r13, %r12                     # a[0]*b[i] = (t0,t2)
+    addq    %r13, $_acc                          # acc += t0
+    movq    %r12, %r10
+    adcq    \$0, %r10                            # t2 += CF
+
+    movq    $_k0, %r13
+    imulq   $_acc, %r13                          # acc * k0
+    andq    $mask52, %r13                        # yi = (acc * k0) & mask52
+
+    vmovq   %r13, $Yi_xmm
+    vpbroadcastq    $Yi_xmm, $Yi                 # broadcast y[i]
+    movq    $_data_offset($m), %rdx
+    mulx    %r13, %r13, %r12                     # yi * m[0] = (t0,t1)
+    addq    %r13, $_acc                          # acc += t0
+    adcq    %r12, %r10                           # t2 += (t1 + CF)
+
+    shrq    \$52, $_acc
+    salq    \$12, %r10
+    or      %r10, $_acc                          # acc = ((acc >> 52) | (t2 << 12))
+
+    lea -168(%rsp), %rsp
+    {vex} vpmadd52luq `$_data_offset+64*0`($a), $Bi, $_R0
+    {vex} vpmadd52luq `$_data_offset+64*0+32`($a), $Bi, $_R0h
+    {vex} vpmadd52luq `$_data_offset+64*1`($a), $Bi, $_R1
+    {vex} vpmadd52luq `$_data_offset+64*1+32`($a), $Bi, $_R1h
+    {vex} vpmadd52luq `$_data_offset+64*2`($a), $Bi, $_R2
+
+    {vex} vpmadd52luq `$_data_offset+64*0`($m), $Yi, $_R0
+    {vex} vpmadd52luq `$_data_offset+64*0+32`($m), $Yi, $_R0h
+    {vex} vpmadd52luq `$_data_offset+64*1`($m), $Yi, $_R1
+    {vex} vpmadd52luq `$_data_offset+64*1+32`($m), $Yi, $_R1h
+    {vex} vpmadd52luq `$_data_offset+64*2`($m), $Yi, $_R2
```

Comment:
it would be more readable if `$_data_offset+32*1` format was used instead of `$_data_offset+64*0+32`

Reply:
Agree! Revised accordingly in the 2k/3k/4k implementation. Thanks!